flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java

```diff
@@ -0,0 +1,218 @@
+						jobResultFuture.completeExceptionally(throwable);
+					} else {
+						jobResultFuture.complete(ArchivedExecutionGraph.createFromFailedInit(
+							jobGraph,
+							throwable,
+							jobStatus,
+							initializationTimestamp));
+					}
+				}
+			}
+			return null;
+		}));
+	}
+
+	public CompletableFuture<ArchivedExecutionGraph> getResultFuture() {
+		return jobResultFuture;
+	}
+
+	public CompletableFuture<JobDetails> requestJobDetails(Time timeout) {
+		synchronized (lock) {
+			if (isRunning()) {
+				return getJobMasterGateway().thenCompose(jobMasterGateway -> jobMasterGateway.requestJobDetails(
+					timeout));
+			} else {
+				int[] tasksPerState = new int[ExecutionState.values().length];
+				return CompletableFuture.completedFuture(new JobDetails(
+					jobGraph.getJobID(),
+					jobGraph.getName(),
+					initializationTimestamp,
+					0,
+					0,
+					jobStatus,
+					0,
+					tasksPerState,
+					jobGraph.getVerticesAsArray().length));
+			}
+		}
+	}
+
+	public CompletableFuture<Acknowledge> cancel(Time timeout) {
+		synchronized (lock) {
+			if (isRunning()) {
+				return getJobMasterGateway().thenCompose(jobMasterGateway -> jobMasterGateway.cancel(
+					timeout));
+			} else {
+				return cancelInternal();
+			}
+		}
+	}
+
+	private CompletableFuture<Acknowledge> cancelInternal() {
+		Preconditions.checkState(Thread.holdsLock(lock));
+		jobStatus = JobStatus.CANCELLING;
+		this.cancellationFuture = new CompletableFuture<>();
+		return cancellationFuture;
+	}
+
+	public CompletableFuture<JobStatus> requestJobStatus(Time timeout) {
+		synchronized (lock) {
+			if (isRunning()) {
+				return getJobMasterGateway().thenCompose(jobMasterGateway -> jobMasterGateway.requestJobStatus(timeout));
+			} else {
+				return CompletableFuture.completedFuture(jobStatus);
+			}
+		}
+	}
+
+	public boolean isRunning() {
+		// todo revisit definition of isRunning. alternative definition: once gateway is available = when is leader
+		return jobStatus == JobStatus.RUNNING;
+	}
+
+	public CompletableFuture<JobMasterGateway> getJobMasterGateway() {
+		synchronized (lock) {
+			Preconditions.checkState(
+				isRunning(),
+				"JobMaster Gateway is not available during initialization");
+			try {
+				return jobManagerRunnerFuture.thenCompose(JobManagerRunner::getJobMasterGateway);
+			} catch (Throwable e) {
+				throw new IllegalStateException("JobMaster gateway is not available", e);
+			}
+		}
+	}
+
+	@Override
+	public CompletableFuture<Void> closeAsync() {
+		synchronized (lock) {
+			switch (jobStatus) {
+				case RUNNING:
+					return jobManagerRunnerFuture.thenAccept(AutoCloseableAsync::closeAsync);
+				case CANCELED:
+				case FAILED:
+					return CompletableFuture.completedFuture(null);
+				case INITIALIZING:
+					return cancelInternal().thenApply(ack -> null);
+				default:
+					throw new IllegalStateException(
+						"Job is in a status that is not covered by the DispatcherJob: "
+							+ jobStatus);
+			}
```

Comment:
What I like to do in `closeAsync` is to always return the same `terminationFuture`.

Reply:
There is a problem if we call `closeAsync` on the `JobStatus.CANCELLING` state.